### PR TITLE
Feature/email confirmation token

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: app/src/app/user/profile/profile.ts [hunk 1]
```diff
@@ -7,10 +7,12 @@ module app.user.profile {
...
 
         static $inject = ['stateHelperServiceProvider'];
-        constructor(private stateHelperServiceProvider){
+        constructor(
+            private stateHelperServiceProvider
+        ){
 
             let state:global.IState = {
-                url: '/profile?passwordResetToken',
+                url: '/profile?passwordResetToken&emailConfirmationToken',
```
Comment (zakhenry) [inline app/src/app/user/profile/profile.ts]: I don't think password reset is needed any more in this, due to the updates with the `authService`

File: app/src/common/services/userService.spec.ts [hunk 2]
```diff
@@ -187,6 +187,28 @@
...
+            it('should be able to send a patch request to confirm email change', () => {
+
+                let user = _.clone(fixtures.user);
+
+                const emailToken = 'cf8a43a2646fd46c2081960ff1150a6b48d5ed062da3d59559af5030eea21548';
+
+                $httpBackend.expectPATCH('/api/users/' + user.userId, user,
+                    (headers) => {
+                        return headers['email-confirm-token'] == emailToken;
+                    }).respond(202);
```
Comment (zakhenry) [inline app/src/common/services/userService.spec.ts:207]: This test should also assert that the `emailConfirmed` key is present in the patch body.

File: app/src/common/services/userService.ts [hunk 3]
```diff
@@ -119,6 +119,18 @@ module common.services.user {
...
         }
+
+        /**
+         * Confirm email update for a user
+         * @param user
+         * @param emailConfirmToken
+         * @returns {ng.IHttpPromise<any>}
+         */
+        public confirmEmail(user:common.models.User, emailConfirmToken:string):ng.IPromise<any>{
+            user['emailConfirmed'] = moment().toISOString();
```
Comment (zakhenry) [inline app/src/common/services/userService.ts]: use dot notation for javascript

File: app/src/common/services/userService.ts [hunk 4]
```diff
@@ -119,6 +119,18 @@ module common.services.user {
...
+            return this.ngRestAdapter
+                .patch('/users/' + user.userId, user, {'email-confirm-token':emailConfirmToken});
```
Comment (zakhenry) [inline app/src/common/services/userService.ts]: The whole user object shouldn't be sent, only the `emailConfirmed` key. This is just for performance/bandwidth reasons, it would work otherwise. Just do `_.pick(user, 'emailConfirmed')`

File: app/src/app/user/profile/profile.ts [hunk 5]
```diff
@@ -21,7 +23,31 @@ module app.user.profile {
...
                 resolve: /*@ngInject*/{
-
+                    emailConfirmationToken:(
+                        userService:common.services.user.UserService,
+                        $stateParams:IStateParams,
+                        ngJwtAuthService:NgJwtAuth.NgJwtAuthService,
+                        $mdToast:ng.material.IToastService
+                    ) => {
+                        if(!_.isEmpty($stateParams['emailConfirmationToken'])) {
+                            userService.confirmEmail(<common.models.User>ngJwtAuthService.getUser(), $stateParams['emailConfirmationToken'])
```
Comment (zakhenry) [inline app/src/app/user/profile/profile.ts]: Use dot notation here

File: api/app/Http/Controllers/UserController.php [hunk 6]
```diff
@@ -116,6 +116,7 @@ public function putOne($id, Request $request)
 
         $model = $this->repository->getNewModel();
         $model->fill($request->all());
+        $model->email = $request->get('email');
```
Comment (zakhenry) [inline api/app/Http/Controllers/UserController.php]: Why is this line needed? Doesn't the `->fill()` method assign the requested email to the model? If the request doesn't contain email this may throw an error?

File: api/app/Http/Controllers/UserController.php [hunk 7]
```diff
@@ -138,23 +139,25 @@ public function patchOne($id, Request $request)
...
 
+        // Check if the email is being changed, and initialize confirmation
+        $email = $request->get('email');
+        if ($email && $model->email != $email) {
+            $token = $model->makeConfirmationToken($email, $this->cache);
+            $this->dispatch(new SendEmailConfirmationEmail($model, $email, $token));
+            $request->merge(['email_confirmed' => null]);
+        }
+
         if ($request->get('email_confirmed')) {
```
Comment (xiphiaz) [inline api/app/Http/Controllers/UserController.php]: This should check that the value hasn't changed as the client might send it anyway when the client intends to update another field, and the email token won't be in cache.

File: api/app/Http/Controllers/UserController.php [hunk 8]
```diff
@@ -138,23 +139,25 @@ public function patchOne($id, Request $request)
...
             $token = $request->headers->get('email-confirm-token');
             if (!$email = $this->cache->pull('email_confirmation_'.$token)) {
                 throw new ValidationException(
                     new MessageBag(['email_confirmed' => 'The email confirmation token is not valid.'])
                 );
+            } else {
```
Comment (zakhenry) [inline api/app/Http/Controllers/UserController.php:155]: This should check that the value hasn't changed as the client might send it anyway when the client intends to update another field, and the email token won't be in cache.
